### DISCUSSION TITLE
Connect with self-hosted Gitlab?

### DISCUSSION THREAD

**Discussion by michaelseidel90** [vercel/next.js#15878] [Help] [answered]
Hi guys,

just wanted to know if there is an option to connect Vercel to a self-hosted Gitlab rep?

Kind regards

**Comment by paulogdm** [comment]
Hello Michael,

Not yet. The team is working on it. However, feel free to use the CLI with GitLab pipelines if you want to: https://vercel.com/knowledge/using-vercel-cli-for-custom-workflows

**Reply by jokinol** [reply]
Any news regarding this? Would help us a lot.

**Reply by maxstoyanov** [reply]
I would like to see support for self hosted GitLab aswell. Just a quick note: Netlify also supports this but chose to make it a higher paid option. Probably because they thought that normal people just use the public instance and only big corporations choose the self hosted instance.

I argue that the opposite is true. I run a little GitLab server for 5 euros per month and use it for all my website projects. Just to use Vercel I started to migrate some projects to the public GitLab instance. I use my self hosted instance for reasons of pricing (started using GitLab during a time where GitHub's private repos were expensive) and autonomy. Both are not very enterprisey. ;)

So, please don't go down the route of Netlify in this case.

**Reply by paulogdm** [reply]
@maxstoyanov Feedback received and shared with our team. Thank you very much for taking the time to write it.

**Reply by dekristo** [reply]
Hi,

I tried your solution but it seems that we need to create a project with github or gitlab before being able to use the vercel cli. Is there any way to create a project by the API?

**Reply by SMTC-Schmidt** [reply]
Hi,

i totally agree with @maxstoyanov.

Is there any news on this topic?
Would be a great feature! :-)

**Reply by snikch** [reply]
I have the same issue as @dekristo. You cannot create a project without using either a public Git repo, or hosted Bitbucket, Github or Gitlab. You literally cannot use Vercel with just a self hosted Gitlab. I'm somewhat astonished and really disappointed. Does anyone have a workaround for this?

**Reply by paulogdm** [reply]
The workaround is to use the CLI: `npm i -g vercel`. You can use `vercel --prod`, for example.

Reach out to our support staff for further info: https://vercel.com/support

**Reply by snikch** [reply]
Thanks @paulogdm. The issue is actually creating a project without a hosted git provider in the first place, which you need *before* you can use the cli tool.

**Reply by BuiltByWalsh** [reply]
@paulogdm Im trying out this solution too, and the vercel CLI doesnt appear to be enough. When using the CLI it expects you to set up a `PROJECT_ID` env variable. But I cannot instantiate a new project in Vercel without linking it to a GitLab or GitHub repo. So it's a chicken and the egg type scenario. Wanna use the CLI, you need a project. Want to create a project? You can't have a self-hosted GitLab.

**Reply by paulogdm** [reply]
@jimboslicethat you can use `vercel` to create it, no? Also, a quick way can be `vercel --confirm`.

**Reply by dekristo** [reply]
@paulogdm it is not possible to use vercel cli to create projects. You have to create beforehand a project on public gitlab or github. Also with the cli tool your are very limited, you cannot for example use multiple environments. Again a workaround could be to create one project by environment :( but very dirty solution. If there is a way to help the team with this topic I volunteer

**Comment by jhilden** [comment]
Is there any update on when support for self-hosted Gitlab will be added to Vercel?

**Comment by hachi-studio** [comment]
Also wondering if there is any update on self hosted Gitlab, it would make things so much easier for my team right now.

**Comment by ericdrgn** [comment]
I'll throw my support into wanting this to happen as well! As mentioned before Netlify charging an arm and a leg for it is pretty ridiculous. I think especially with the current climate of everything in the world getting this working would be awesome for you. You would be drawing in a lot of open source projects and people working with that themselves. Personally I can say I have around 20 projects in Netlify right now that (using github sadly) that I would quickly switch over to Vercel if you guys supported my self hosted gitlab. @paulogdm

**Reply by paulogdm** [reply]
I relayed your comment to the team! Thanks for sharing your thoughts.

**Reply by Irrelon** [reply]
Agreed. I also run my own gitlab server because github made private repos limited and expensive to make unlimited. I was surprised that I couldn't get self-hosted gitlab to auto-connect with vercel.

**Comment by danielweil** [comment]
Is there any updates on this request? Thanks!

**Comment by SinaKarimi7** [comment]
any update?!

**Comment by yordis** [comment]
Hey folks, we are facing some pain points with Vercel since we have a self-hosted GitLab, I would appreciate some insightful updates from the Vercel team, and/or share some insights into the potential work required to make it work.

I am more than happy to contribute if some of the work required is open-source since it would alleviate a lot of pain from our side.

**Reply by samcx** [reply]
Hi there,

Unfortunately, no updates on this yet, but it is on our roadmap! I have let our Product team know.

**Reply by yordis** [reply]
Would be too much to ask for a time?

The comment https://github.com/vercel/next.js/discussions/15878#discussioncomment-47123 was on Aug 5, 2020  "Not yet. The team is working on it." 

So I am a bit worry it keeps being kick further and further, has been long enough 😭  I am offering to help if there is anything I can do, but it requires some direction and openness from you.

I would appreciate some transparency on the topic 🙏🏻

**Reply by samcx** [reply]
Hi there,

There is, unfortunately, no timetable for this yet — our Product/Engineering team only has so much bandwidth while having to juggle with other features that are of varying importance. Our team is growing though! Hopefully we'll have the capacity to tackle this soon. Not the first time I've seen this request, so each new request for this feature will only increase our priority for getting it done 🙌🏼

**Reply by napalmj** [reply]
Would love to see this option as well. It's definitely a deciding factor for my company when using Vercel vs Netlify. Would love to use Vercel because it works better with Nextjs but sadly this is not an option.

**Reply by zbruhnke** [reply]
Just piling on here - ultimately "supporting" self hosted gitlab should simply be letting us put in our own gitlab base URL. The API is identical and if things break you can very likely tell us we should probably update our gitlab to the current version because if the hosted version is working the self hosted version should be working too

**Comment by karlomedallo** [comment]
If you guys are still interested in this feature, please upvote! https://github.com/vercel/community/discussions/337

**Comment by leerob** [comment, marked as answer]
**Update:** It's now much easier to deploy to Vercel with any custom CI/CD provider like self-managed GitLab with GitLab Pipelines!

[We've created a guide](https://vercel.com/support/articles/how-can-i-use-gitlab-pipelines-with-vercel) and [an example](https://github.com/vercel/examples/tree/main/ci-cd/gitlab-cicd) using GitLab Pipelines. Hope this helps! 🙏

**Reply by TheCodeDestroyer** [reply]
@leerob what if we want to give Vercel access to our self managed git repo as it might be a monorepo which Vercel handles really well, do you not support that?
